assets/js/index.js

```diff
@@ -0,0 +1,97 @@
+let library = setupLibrary();
+
+if (!localStorage.uniqueId) {
+  localStorage.uniqueId = 1; 
+}
+
+function setupLibrary() {
+  let library;
+  if(localStorage.library) {
+    library = JSON.parse(localStorage.library).map((e) => Object.assign(new Book(), e));
```

Comment:
Can you help me understand the logic behind this piece of code:
```javascript
Object.assign(new Book(), e)
```
I do understand what `Object.assign` does. What I wanna know is why we are instantiating a `Book` object here. From how I understand your `setupLibrary` function, you are retrieving the array stored in the `localStorage` or return an empty array (suppose localStorage is empty). If so, why then didn't we simply do this:
```javascript
library = JSON.parse(localStorage.library)
``` ?

Reply:
Okay, this is interesting. It looks we don't need to use type casting in here. But if we want to handle instance level operations(in our case it's toggleRead), I'm gonna use the way we did with type casting. If we don't use casting, we needed to handle toggleRead(there will be a lot of instance methods in some future) manually. And you don't want to handle all these kind of methods via the wrapper.

Reply:
If we just recover it from the `localStorage` without type casting, the books are not gonna be books. They're gonna be objects, ok, but not instances of the `Book` class and we're not gonna be able to call `toggleRead()` on them. Without the type casting, we would have a big problem to handle the read status changing of old books after a page refresh.

Reply:
Javascript isn't a pure object-oriented language. But in here, we're trying to build a project that has so many elements that control business level things such as interacting with the persistence layer. For providing the multi-layer application architecture, we need to divide our components by its responsibilities, through we need to use objects and it's behaviors for managing this separation. All the high-level object-oriented languages support type-casting(java, c#), but javascript ruby and python doesn't have strong types. It has some benefits and harms. But if you want to use javascript as like the super-object-oriented languages(java, c#), you always need to use type checking and type-casting, btw javascript community built typescript for this reason.

Reply:
Thank you @btuerker and @madcido for sharing. I appreciate the learning I got here. Ok now I understand (not the details of your architecture but the general idea)! :)